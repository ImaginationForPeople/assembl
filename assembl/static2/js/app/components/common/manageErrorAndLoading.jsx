// @flow
/*
  Higher order component that:
  - if there is a graphql error throws it (this error will be handled by an error boundary component)
  - if loading is true displays the loader component or nothing
  - Can pass custom Loader component to override the default
  - if there is no error and loading is false displays the wrapped component
*/
import * as React from 'react';
import Loader, { type Props as LoaderProps } from './loader';
import WatsonLoader, { LOADER_TYPE as LOADER_TYPE_WATSON } from './loader/loader';

type Props = {
  displayLoader: boolean,
  /** Optional type */
  loaderType?: string
} & LoaderProps;

type WrappedProps = {
  data: { error?: ?Error, loading?: boolean } & any,
  error?: ?Error,
  loading?: boolean
};

// Takes the type and return the right loader for error and loading
const switchLoaderToShow = (loaderType?: string, propsToPass: Props) => {
  const type = loaderType || '';
  switch (type) {
  case 'watson': {
    return {
      ERROR: () => <WatsonLoader type={LOADER_TYPE_WATSON.ERROR} />,
      LOADING: () => <WatsonLoader type={LOADER_TYPE_WATSON.LOADING} />
    };
  }
  default: {
    return {
      ERROR: () => {
        throw new Error('GraphQL error');
      },
      LOADING: () => <Loader {...propsToPass} />
    };
  }
  }
};

const manageErrorAndLoading = (props: Props) => (WrappedComponent: React.ComponentType<any>) => (wrappedProps: WrappedProps) => {
  const { data, error, loading } = wrappedProps;
  const { loaderType } = props;

  const loaderToShow = switchLoaderToShow(loaderType, props);

  // ERROR
  if (error || (data && data.error)) {
    const graphqlError = error || data.error;
    if (graphqlError) {
      loaderToShow.ERROR();
    }
  }
  // LOADING
  if (loading || (data && data.loading)) {
    return props.displayLoader ? loaderToShow.LOADING() : null;
  }
  // WRAPPED COMPONENT
  return <WrappedComponent {...wrappedProps} />;
};

manageErrorAndLoading.defaultProps = {
  loaderType: 'default'
};

export default manageErrorAndLoading;