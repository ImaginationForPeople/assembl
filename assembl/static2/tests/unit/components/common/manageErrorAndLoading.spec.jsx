// @flow
import React from 'react';
import renderer from 'react-test-renderer';

import manageErrorAndLoading, { TYPE } from '../../../../js/app/components/common/manageErrorAndLoading';

const DummyComponent = () => <div>My dummy component</div>;

describe('manageErrorAndLoading HOC', () => {
  it('should throw an error if error is not null/undefined', () => {
    const props = { displayLoader: true };
    const wrappedProps = {
      data: {},
      error: new Error('My awesome error'),
      loading: false
    };
    expect(() => manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toThrowError('GraphQL error');
  });

  it('should throw an error if error is not null/undefined', () => {
    const props = { displayLoader: true };
    const wrappedProps = {
      data: {},
      error: new Error('My awesome error'),
      loading: false
    };
    expect(() => manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toThrowError('GraphQL error');
  });

  it('should render an error component if data.error or error is not null/undefined and type is watson', () => {
    const props = { displayLoader: true, loaderType: TYPE.SEMANTIC_ANALYSIS };
    const wrappedProps = {
      data: {
        error: new Error('My graphql error')
      }
    };
    const rendered = renderer.create(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should return null if data are loading and we don\'t want to display a loader (use data prop)', () => {
    const props = { displayLoader: false };
    const wrappedProps = {
      data: {
        error: null,
        loading: true
      }
    };
    expect(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toBeNull();
  });

  it('should return null if data are loading and we don\'t want to display a loader', () => {
    const props = { displayLoader: false };
    const wrappedProps = {
      data: {},
      error: null,
      loading: true
    };
    expect(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toBeNull();
  });

  it('should display a loader if data are loading and we want to display a loader (use data prop)', () => {
    const props = { displayLoader: true };
    const wrappedProps = {
      data: {
        error: null,
        loading: true
      }
    };
    const rendered = renderer.create(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should display a loader if data are loading and we want to display a loader', () => {
    const props = { displayLoader: true };
    const wrappedProps = {
      data: {},
      error: null,
      loading: true
    };
    const rendered = renderer.create(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should render the component if everything is OK (use data prop)', () => {
    const props = { displayLoader: true };
    const wrappedProps = {
      data: {
        error: null,
        loading: false
      }
    };
    const rendered = renderer.create(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should render the component if everything is OK', () => {
    const props = { displayLoader: true };
    const wrappedProps = {
      data: {},
      error: null,
      loading: false
    };
    const rendered = renderer.create(manageErrorAndLoading(props)(DummyComponent)(wrappedProps)).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});