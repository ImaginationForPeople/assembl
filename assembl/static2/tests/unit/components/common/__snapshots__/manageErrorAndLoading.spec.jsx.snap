// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manageErrorAndLoading HOC should display a loader if data are loading and we want to display a loader (use data prop) 1`] = `
<div
  className="loader-container-xl"
>
  <div
    className="loader"
  >
    <span>
      Please wait...
    </span>
    <div
      className="relative"
    >
      <svg
        height="30"
        viewBox="0 0 120 30"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="15"
          cy="15"
          r="15"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="15"
            repeatCount="indefinite"
            to="15"
            values="15;9;15"
          />
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="1"
            repeatCount="indefinite"
            to="1"
            values="1;.5;1"
          />
        </circle>
        <circle
          cx="60"
          cy="15"
          fillOpacity="0.3"
          r="9"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="9"
            repeatCount="indefinite"
            to="9"
            values="9;15;9"
          />
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="0.5"
            repeatCount="indefinite"
            to="0.5"
            values=".5;1;.5"
          />
        </circle>
        <circle
          cx="105"
          cy="15"
          r="15"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="15"
            repeatCount="indefinite"
            to="15"
            values="15;9;15"
          />
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="1"
            repeatCount="indefinite"
            to="1"
            values="1;.5;1"
          />
        </circle>
      </svg>
    </div>
  </div>
</div>
`;

exports[`manageErrorAndLoading HOC should display a loader if data are loading and we want to display a loader 1`] = `
<div
  className="loader-container-xl"
>
  <div
    className="loader"
  >
    <span>
      Please wait...
    </span>
    <div
      className="relative"
    >
      <svg
        height="30"
        viewBox="0 0 120 30"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="15"
          cy="15"
          r="15"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="15"
            repeatCount="indefinite"
            to="15"
            values="15;9;15"
          />
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="1"
            repeatCount="indefinite"
            to="1"
            values="1;.5;1"
          />
        </circle>
        <circle
          cx="60"
          cy="15"
          fillOpacity="0.3"
          r="9"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="9"
            repeatCount="indefinite"
            to="9"
            values="9;15;9"
          />
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="0.5"
            repeatCount="indefinite"
            to="0.5"
            values=".5;1;.5"
          />
        </circle>
        <circle
          cx="105"
          cy="15"
          r="15"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="15"
            repeatCount="indefinite"
            to="15"
            values="15;9;15"
          />
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="1"
            repeatCount="indefinite"
            to="1"
            values="1;.5;1"
          />
        </circle>
      </svg>
    </div>
  </div>
</div>
`;

exports[`manageErrorAndLoading HOC should render an error component if data.error or error is not null/undefined and type is watson 1`] = `
<div>
  My dummy component
</div>
`;

exports[`manageErrorAndLoading HOC should render the component if everything is OK (use data prop) 1`] = `
<div>
  My dummy component
</div>
`;

exports[`manageErrorAndLoading HOC should render the component if everything is OK 1`] = `
<div>
  My dummy component
</div>
`;
